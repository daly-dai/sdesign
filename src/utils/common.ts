/**
 * 生成唯一code
 * @param num code的位数
 * @returns
 */
export function createCode(num = 4) {
  const UP_ALPHABET = [
    'A',
    'B',
    'C',
    'D',
    'E',
    'F',
    'G',
    'H',
    'I',
    'J',
    'K',
    'L',
    'M',
    'N',
    'O',
    'P',
    'Q',
    'R',
    'S',
    'T',
    'U',
    'V',
    'W',
    'X',
    'Y',
    'Z',
  ];
  const DOWN_ALPHABET = [
    'a',
    'b',
    'c',
    'd',
    'e',
    'f',
    'g',
    'h',
    'i',
    'j',
    'k',
    'l',
    'm',
    'n',
    'o',
    'p',
    'q',
    'r',
    's',
    't',
    'u',
    'v',
    'w',
    'x',
    'y',
    'z',
  ];
  const NUMBER = ['1', '2', '3', '4', '5', '6', '7', '8', '9']; //用数组chars保存验证码里出现的字符

  const chars = UP_ALPHABET.concat(DOWN_ALPHABET, NUMBER);

  let randCode = ''; //定义一个初始值为空的字符串变量为最终产生的随机验证码

  for (let i = 0; i < num; i++) {
    //0-1的随机小数 -->  0~数组长度-1的范围   取整
    const randPosition = Math.floor(Math.random() * (chars.length - 1)); //每次生成一个随机数的位置

    randCode += chars[randPosition]; //带有随机位置作为下标，指示到当前随机产生的某个字符上，将其连接到随机验证码的后面
  }

  return randCode;
}

/**
 * 处理文件名称,超过限制字数保留前三后三
 * @param name
 * @returns
 */
export function dispatchFileName(name: string, nameLimit: number) {
  if (!name) return '';

  let FileDataType = '';

  if (name.includes('.')) {
    const nameList = name.split('.');

    if (nameList?.length > 1) {
      FileDataType = `.${nameList[nameList.length - 1]}`;
    }
  }

  let fileName = name;

  if (FileDataType) {
    fileName = name.replace(FileDataType, '');
  }

  if (fileName.length > nameLimit) {
    const lastThreeChars = fileName.substring(fileName.length - 3); // 截取后三个字符
    const firstThreeChars = fileName.substring(0, 3); // 截取前三个字符

    return `${firstThreeChars}...${lastThreeChars}${FileDataType}`;
  }

  return name;
}

/**
 * @description 创建数组
 * @param num
 * @returns
 */
export function createArray(num = 4) {
  return Array.from({ length: num }, (_, i) => i);
}

/**
 * 获取文件base64相关数据
 * @param file
 * @returns
 */
export function getBase64(file: any): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => resolve(reader.result as string);
    reader.onerror = (error) => reject(error);
  });
}
/**
 * 动态引入js
 *
 * @param {string} src - js链接地址
 * @example
 *
 * ```ts
 * import { injectScript } from 'super-tools-lib'
 *
 * injectScript(src)
 * ```
 */
export function injectScript(src: string) {
  const s = document.createElement('script');
  s.type = 'text/javascript';
  s.async = true;
  s.src = src;
  const t = document.getElementsByTagName('script')[0] as any;

  t.parentNode.insertBefore(s, t);
}

/**
 * 复制文本到剪切板
 * @param text 需要复制的文本
 */
export async function copyToClipboard(
  text: string,
  successCb?: () => void,
  errorCb?: () => void,
): Promise<void> {
  // 检查权限的示例函数，具体实现取决于你的框架和API
  async function checkPermission(): Promise<void> {
    if (!('clipboardWrite' in navigator.clipboard)) {
      throw new Error('Clipboard API is not supported.');
    }
  }

  try {
    // 检查是否具有剪贴板写入权限的函数
    await checkPermission(); // 检查权限，根据实际情况可能需要不同的实现
    await navigator.clipboard.writeText(text);
    successCb?.();
  } catch (error) {
    console.error('Failed to copy text to clipboard:', error);
    errorCb?.();
  }
}

const isNode =
  typeof process !== 'undefined' &&
  process.versions !== null &&
  process.versions.node !== null;

/**
 * 用于判断当前是否在浏览器环境中。
 * 首先会判断当前是否处于测试环境中（通过 process.env.NODE_ENV === 'TEST' 判断），
 * 如果是，则返回 true。否则，会进一步判断是否存在 window 对象、document 对象以及 matchMedia 方法
 * 同时通过 !isNode 判断当前不是在服务器（Node.js）环境下执行，
 * 如果都符合，则返回 true 表示当前处于浏览器环境中。
 * @returns  boolean
 */
export const isBrowser = () => {
  if (typeof process !== 'undefined' && process.env.NODE_ENV === 'TEST') {
    return true;
  }
  return (
    typeof window !== 'undefined' &&
    typeof window.document !== 'undefined' &&
    typeof window.matchMedia !== 'undefined' &&
    !isNode
  );
};
