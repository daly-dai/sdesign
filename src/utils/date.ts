import dayjs, { Dayjs } from 'dayjs';

import { RangeValueType } from '../components';

function isDayjsInstance(obj: dayjs.Dayjs) {
  // 检查对象是否存在并且不是原始类型
  if (obj && typeof obj === 'object' && !Array.isArray(obj)) {
    // 检查对象是否有 Day.js 常见的属性或方法
    // 注意：这里只是一个示例，Day.js 的 API 可能会变化
    // 你可以根据 Day.js 的当前版本和你想要检查的具体属性/方法进行修改
    return (
      typeof obj.format === 'function' && // Day.js 实例有 format 方法
      typeof obj.toDate === 'function' && // Day.js 实例可以转换为 Date 对象
      // ... 可以添加更多 Day.js 特有的检查
      true // 如果上述条件都满足，则假设它是 Day.js 实例
    );
  }
  return false; // 如果不是对象，或者没有上述属性/方法，则不是 Day.js 实例
}

export function getDateVal(data: RangeValueType | undefined) {
  if (!data) return null;

  if (typeof data === 'string') {
    if (!dayjs(data)?.isValid()) return null;

    return dayjs(data);
  }

  if (Array.isArray(data) && data?.length) {
    return data?.map((item: string | null) => {
      if (!dayjs(item)?.isValid()) return null;

      return dayjs(item);
    });
  }

  if (typeof data === 'object' && isDayjsInstance(data as Dayjs)) {
    return data;
  }

  return null;
}
