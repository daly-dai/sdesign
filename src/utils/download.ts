/**
 * 下载二进制流
 * @param res
 * @param name
 */
export function downloadBlob(res: any, name: string) {
  const blob = new Blob([res.data]);

  const anchor = document.createElement('a');
  const href = window.URL.createObjectURL(blob); //关键点3
  anchor.setAttribute('href', href);
  // /* 关键之处：使用download属性必须要html5的页面才行 ，而且它不会刷新，文件名及扩展名均由这里控制*/
  anchor.setAttribute('download', name); //关键点4
  anchor.click();
}

/**
 * 下载链接
 * @param href
 * @param title
 */
export function downloadLink(href: string, title?: string) {
  let findTitle: string = '';

  if (!title) {
    const list = href.split('/');
    if (list.length > 0) {
      findTitle = list[list.length - 1];
    }
  }

  const a = document.createElement('a');
  a.setAttribute('href', href);
  a.setAttribute('download', title || findTitle);
  a.click();
}
