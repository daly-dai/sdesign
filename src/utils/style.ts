import { isString } from 'lodash';
import isNumber from 'lodash/isNumber';

import { isStringNumber } from './data-types';
import { debugWarn } from './error';

const SCOPE = 'utils/style';

export function addUnit(value?: string | number, defaultUnit = 'px') {
  if (!value) return '';

  if (isNumber(value) || isStringNumber(value)) {
    return `${value}${defaultUnit}`;
  }

  if (isString(value)) {
    return value;
  }

  debugWarn(SCOPE, 'binding value must be a string or number');
}
