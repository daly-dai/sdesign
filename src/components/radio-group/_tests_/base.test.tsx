import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import { describe, expect, it, vi } from 'vitest';

import SRadioGroup from '../index';

import SConfigProvider from '@daly/sdesign/components/config-provider';
import mountTest from '@daly/sdesign/tests/mountTest';
import { waitFakeTimer } from '@daly/sdesign/tests/utils';

const globalDict = {
  level: {
    L1: '不敏感',
    L2: '低敏感',
    L3: '较敏感',
    L4: '敏感',
    L5: '极敏感',
  },
  type: {
    app_link: '存量应用关联',
    register: '发布注册',
  },
};

describe('SRadioGroup', () => {
  mountTest(SRadioGroup);
  mountTest(() => <SRadioGroup options={[]} />);
  it('should render correctly', () => {
    const { container } = render(<SRadioGroup />);
    expect(container).toMatchSnapshot();
  });

  it('should render dictKey', () => {
    const { container, getByText } = render(
      <SConfigProvider globalDict={globalDict}>
        <SRadioGroup dictKey="level" />
      </SConfigProvider>,
    );

    expect(container).toMatchSnapshot();

    waitFakeTimer();

    expect(getByText('不敏感')).toBeInTheDocument();
  });

  it('should options props priority dict props', () => {
    const onChange = vi.fn();

    const { container, getByText } = render(
      <SRadioGroup
        dict={globalDict.level}
        onChange={onChange}
        options={[
          {
            label: '存量应用关联',
            value: 'app_link',
          },
          {
            label: '发布注册',
            value: 'register',
          },
        ]}
      />,
    );

    expect(container).toMatchSnapshot();

    waitFakeTimer();

    expect(getByText('存量应用关联')).toBeInTheDocument();
  });

  it('should work basically', () => {
    const onChange = vi.fn();

    const { container } = render(
      <SRadioGroup dict={globalDict.level} onChange={onChange} />,
    );

    waitFakeTimer();

    fireEvent.click(container.querySelectorAll('.ant-radio-wrapper')[0]);
    expect(onChange).toHaveBeenCalledWith('L1');
    fireEvent.click(container.querySelectorAll('.ant-radio-wrapper')[1]);
    expect(onChange).toHaveBeenCalledWith('L2');
    fireEvent.click(container.querySelectorAll('.ant-radio-wrapper')[2]);
    expect(onChange).toHaveBeenCalledWith('L3');
    fireEvent.click(container.querySelectorAll('.ant-radio-wrapper')[1]);
    expect(onChange).toHaveBeenCalledWith('L1');
  });

  it('should support disableKeys', () => {
    const onChangeGroup = vi.fn();

    const { container } = render(
      <SConfigProvider globalDict={globalDict}>
        <SRadioGroup dictKey="level" disableKeys={['L1']} />
      </SConfigProvider>,
    );

    fireEvent.click(
      container.querySelectorAll('.ant-radio-wrapper-disabled')[0],
    );
    expect(onChangeGroup).not.toHaveBeenCalled();
  });
});
