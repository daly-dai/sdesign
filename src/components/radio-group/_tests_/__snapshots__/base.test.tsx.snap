// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SRadioGroup > should options props priority dict props 1`] = `
<div>
  <div
    class="ant-radio-group ant-radio-group-outline css-dev-only-do-not-override-3rel02"
  >
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-3rel02"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="app_link"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        存量应用关联
      </span>
    </label>
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-3rel02"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="register"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        发布注册
      </span>
    </label>
  </div>
</div>
`;

exports[`SRadioGroup > should render correctly 1`] = `
<div>
  <div
    class="ant-radio-group ant-radio-group-outline css-dev-only-do-not-override-3rel02"
  />
</div>
`;

exports[`SRadioGroup > should render dictKey 1`] = `
<div>
  <div
    class="ant-radio-group ant-radio-group-outline css-dev-only-do-not-override-3rel02"
  >
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-3rel02"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="L1"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        不敏感
      </span>
    </label>
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-3rel02"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="L2"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        低敏感
      </span>
    </label>
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-3rel02"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="L3"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        较敏感
      </span>
    </label>
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-3rel02"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="L4"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        敏感
      </span>
    </label>
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-3rel02"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="L5"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        极敏感
      </span>
    </label>
  </div>
</div>
`;

exports[`SRadioGroup should options props priority dict props 1`] = `
<div>
  <div
    class="ant-radio-group ant-radio-group-outline css-dev-only-do-not-override-1r287do"
  >
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-1r287do"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="app_link"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        存量应用关联
      </span>
    </label>
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-1r287do"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="register"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        发布注册
      </span>
    </label>
  </div>
</div>
`;

exports[`SRadioGroup should render correctly 1`] = `
<div>
  <div
    class="ant-radio-group ant-radio-group-outline css-dev-only-do-not-override-1r287do"
  />
</div>
`;

exports[`SRadioGroup should render dictKey 1`] = `
<div>
  <div
    class="ant-radio-group ant-radio-group-outline css-dev-only-do-not-override-1r287do"
  >
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-1r287do"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="L1"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        不敏感
      </span>
    </label>
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-1r287do"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="L2"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        低敏感
      </span>
    </label>
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-1r287do"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="L3"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        较敏感
      </span>
    </label>
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-1r287do"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="L4"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        敏感
      </span>
    </label>
    <label
      class="ant-radio-wrapper css-dev-only-do-not-override-1r287do"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="L5"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        极敏感
      </span>
    </label>
  </div>
</div>
`;
